compiler/rustc_lint/src/builtin.rs

```diff
@@ -1803,39 +1761,39 @@ declare_lint! {
     report_in_external_macro
 }
 
-pub struct UnnameableTestItems {
-    boundary: Option<hir::OwnerId>, // Id of the item under which things are not nameable
-    items_nameable: bool,
+#[derive(Default)]
+pub(crate) struct UnnameableTestItems {
+    current_module_is_nameable: Option<bool>,
 }
 
 impl_lint_pass!(UnnameableTestItems => [UNNAMEABLE_TEST_ITEMS]);
 
-impl UnnameableTestItems {
-    pub fn new() -> Self {
-        Self { boundary: None, items_nameable: true }
-    }
-}
-
 impl<'tcx> LateLintPass<'tcx> for UnnameableTestItems {
-    fn check_item(&mut self, cx: &LateContext<'_>, it: &hir::Item<'_>) {
-        if self.items_nameable {
-            if let hir::ItemKind::Mod(..) = it.kind {
-            } else {
-                self.items_nameable = false;
-                self.boundary = Some(it.owner_id);
+    fn check_mod(&mut self, cx: &LateContext<'_>, _: &hir::Mod<'tcx>, n: hir::HirId) {
+        let mut module_is_nameable = true;
+
+        let mut def_id = n.expect_owner().def_id;
+        while let Some(parent_mod) = cx.tcx.opt_local_parent(def_id) {
+            if cx.tcx.def_kind(parent_mod) != DefKind::Mod {
+                module_is_nameable = false;
+                break;
             }
-            return;
-        }
 
-        let attrs = cx.tcx.hir().attrs(it.hir_id());
-        if let Some(attr) = attr::find_by_name(attrs, sym::rustc_test_marker) {
-            cx.emit_spanned_lint(UNNAMEABLE_TEST_ITEMS, attr.span, BuiltinUnnameableTestItems);
+            def_id = parent_mod;
         }
+
+        self.current_module_is_nameable = Some(module_is_nameable);
     }
 
-    fn check_item_post(&mut self, _cx: &LateContext<'_>, it: &hir::Item<'_>) {
-        if !self.items_nameable && self.boundary == Some(it.owner_id) {
-            self.items_nameable = true;
+    fn check_item(&mut self, cx: &LateContext<'_>, it: &hir::Item<'_>) {
+        let parent_mod = cx.tcx.local_parent(it.owner_id.def_id);
+
+        let item_is_nameable =
+            self.current_module_is_nameable.unwrap() && cx.tcx.def_kind(parent_mod) == DefKind::Mod;
+
+        let Some(attr) = cx.tcx.get_attr(it.owner_id, sym::rustc_test_marker) else { return };
+        if !item_is_nameable {
+            cx.emit_spanned_lint(UNNAMEABLE_TEST_ITEMS, attr.span, BuiltinUnnameableTestItems);
```

Comment:
I see the initial eddyb's response to this lint (https://github.com/rust-lang/rust/issues/36629#issuecomment-404753945) and share the sentiment.

At worst, it should be `buffer_lint`ed somewhere in the implementation of `#[test]` or test harness generation, where the source of truth about skipping test items is.